Fix race between MountVolume and UnmountDevice

jsafrane opened the pull request:
When kubelet receives a new pod while UnmountDevice is in progress, it should not enqueue 
 MountVolume operation - the volume device is being unmounted right now.

In addition, don't clear devicePath after UnmountDevice, because subsequent MountDevice may need it.

/kind bug
/sig storage
Fixes #65246

**Does this PR introduce a user-facing change?**:
<!--  
If no, just write "NONE".
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".
2. 
-->
```release-note
Fixed kubelet reporting "resource name may not be empty" when mounting a volume very quickly after unmount.
```

jsafrane commented:
@msau42 @gnufied @jingxu97 @saad-ali @pohly PTAL.

jsafrane commented:
/priority important-soon

pohly commented on pkg/kubelet/volumemanager/reconciler/reconciler.go:
```diff
@@ -166,6 +166,10 @@ func (rc *reconciler) reconcile() {
 		if !rc.desiredStateOfWorld.PodExistsInVolume(mountedVolume.PodName, mountedVolume.VolumeName) {
+			if rc.operationExecutor.IsOperationPending(mountedVolume.VolumeName, nestedpendingoperations.EmptyUniquePodName) {
+				klog.V(5).Infof("Skipping UnmountVolume, device operation is in progress")
+				continue
```
How is this different from the check that the executer will do internally? See here:
https://github.com/kubernetes/kubernetes/blob/b7e2980a57289a32a7dbc5165584a0b1a56c9dd9/pkg/volume/util/nestedpendingoperations/nestedpendingoperations.go#L94-L108

When I asked about it in https://github.com/kubernetes/kubernetes/issues/70319 and later in a CSI WG meeting, @saad-ali said that the executioner will prevent running multiple operations against the same volume in parallel.

It wasn't obvious how that works and I was still worried that the executioner would merely serialize operations. But after looking at the code above, my interpretation is that it will discard new operations *instead* of queuing them, so it should do something similar to what you propose here.

But I am not entirely sure - perhaps your check is stricter, or I misunderstood something.

jsafrane replied:
The difference is `podName` - (Un)MountDevice operation uses empty pod name, it's global for all pods, while (Un)MountVolume uses pod names. IMO, (volumeName, podName) is the primary key in the operation map, so multiple MountVolumes can run in parallel on the same volume.

jsafrane commented:
/retest

pohly commented:
Jan Šafránek <notifications@github.com> writes:
> The difference is `podName` - (Un)MountDevice operation uses empty pod
> name, it's global for all pods, while (Un)MountVolume uses pod
> names. IMO, (volumeName, podName) is the primary key in the operation
> map, so multiple MountVolumes can run in parallel on the same volume.

Then let me come back to my question from
https://github.com/kubernetes/kubernetes/issues/70319#issuecomment-436792272:
is there there ever a situation where there should be more than one
operation in flight *per volume*? Note that I meant "should be" as in
"should be allowed to be".

@saad-ali answer was no, but in practice it was allowed because the
check was was based on volume + pod. That check has been shown to be
insufficient for one case. Can we be sure that this case (which gets
addressed with your PR) is the only one?

I genuinely don't know.

gnufied commented on pkg/kubelet/volumemanager/cache/actual_state_of_world.go:
```diff
@@ -568,7 +568,9 @@ func (asw *actualStateOfWorld) SetVolumeGloballyMounted(
-	volumeObj.devicePath = devicePath
+	if devicePath != "" {
+		volumeObj.devicePath = devicePath
+	}
```
Do we really need this? If we prevent concurrent operation on same volume - won't a pod that uses same volume will pretty much cause another `MountDevice` operation?

jsafrane replied:
Yes, we need it. Consider UnmountDevice is in progress and a new pod arrives.

1. UnmountDevice finishes. This calls `MarkDeviceAsUnmounted` which clears the devicePath, i.e. calls `SetVolumeGloballyMounted(..., devicePath="", ...)`
2. New MountDevice (WaitForAttach) starts and sees empty devicePath -> problems.

jingxu97 replied:
sorry to put a comment this late. very good catch of this bug. I think to avoid such bugs, it is better to not reuse SetVolumeGloballyMounted for both MarkDeviceAsMounted and MarkDeviceAsUnmounted (this function just to set the bool globallyMounted to false so no need to pass devicePath or deviceMountPath)

jsafrane commented:
> is there there ever a situation where there should be more than one
operation in flight *per volume*? Note that I meant "should be" as in
"should be allowed to be".

I am not sure. However, in this case we did not see two operations in parallel, we saw UnmountDevice running and MountVolume enqueued. MountVolume was started after UnmountDevice finished and *that* is wrong - the device is not mounted, so MountVolume has nothing to bind-mount. We can:

* Either not enqueue MountVolume operation when UnmountDevice is in progress (that's this PR)
* Or update MountVolume operation to check if MountDevice was completed.

pohly commented:
Jan Šafránek <notifications@github.com> writes:

>> is there there ever a situation where there should be more than one
> operation in flight *per volume*? Note that I meant "should be" as in
> "should be allowed to be".
>
> I am not sure. However, in this case we did not see two operations in
> parallel, we saw UnmountDevice running and MountVolume
> enqueued. MountVolume was started after UnmountDevice finished and
> *that* is wrong - the device is not mounted, so MountVolume has
> nothing to bind-mount. We can:
>
> * Either not enqueue MountVolume operation when UnmountDevice is in progress (that's this PR)
> * Or update MountVolume operation to check if MountDevice was
> completed.

Third option:
 * make the check in the executor stricter by only using the volume as
   key when checking for existing operations

The result in this case would have been that the MountVolume operation
would have been discarded by the executor itself, instead of having to
add that logic to the reconciler.

This would also catch other errors where operations are enqueued that
shouldn't be enqueued because the pending operation will change the
state of the world.

I'd expect the reconciler to simply re-create operations that are still
needed once the running operation is done.

But perhaps such a change has negative effects in cases where
(currently) pending or parallel operations are possible and
desirable. <shrug>

jsafrane committed:
Fixed clearing of devicePath after UnmountDevice


UnmountDevice must not clear devicepath, because such devicePath
may come from node.status (e.g. on AWS) and subsequent MountDevice
operation (that may be already enqueued) needs it.

jsafrane commented:
It seems I was wrong. Only one `MountVolume()` operation runs when multiple pods use the same volume, because pod name is not used in the operation name:

https://github.com/kubernetes/kubernetes/blob/954996e231074dc7429f7be1256a579bedd8344c/pkg/volume/util/operationexecutor/operation_executor.go#L742

So operation checks in this PR are useless and only `devicePath` check is needed. I reworked the PR, now it's one `if`.

pohly commented:
Jan Šafránek <notifications@github.com> writes:
> So operation checks in this PR are useless and only `devicePath` check
> is needed. I reworked the PR, now it's one `if`.

Is the effect different from what I have in
https://github.com/kubernetes/kubernetes/pull/70746?

SetVolumeGloballyMounted is a function with poorly defined semantic, and
adding yet another special case doesn't make it better. IMHO deleting it
and moving the relevant code to MarkDeviceAsMounted and
MarkDeviceAsUnmounted is cleaner. Just my 2 cents of course, I don't
care which PR gets merged (if any).

msau42 commented:
Please add unit tests to this function and reconciler like I suggested here: https://github.com/kubernetes/kubernetes/pull/70746#discussion_r233228120

jsafrane commented:
Unit test added. It fails without changes in `actual_state_of_world.go` and passes with them.

gnufied commented:
/lgtm

but I also think that - we should also merge https://github.com/kubernetes/kubernetes/pull/71095 because while we have fixed one race here, not relying on `attachID` value present in actual state of the world, while calling `WaitForAttach` makes sense. Many other drivers as well discover the actual device path in `WaitForAttach` call itself.

AishSundar commented:
@saad-ali @msau42 looks like this is ready for merge. let me know if you want help stamping it with 1.13 milestone

AishSundar commented:
/milestone v1.13
/priority critical-urgent
/remove-priority important-soon

gnufied commented:
Lets cherry-pick this to 1.12 and 1.11. @jsafrane @msau42 any objectios. I think this will also fix long standing bug https://github.com/kubernetes/kubernetes/issues/67342

msau42 commented:
sgtm

dennis-benzinger-hybris commented:
Is there any chance of getting this merge to 1.10 ?

jsafrane commented:
> Is there any chance of getting this merge to 1.10 ?

No, Kubernetes supports only 3 releases at time (1.11 - 1.13).

dennis-benzinger-hybris commented:
Thanks @jsafrane. That confirms what I expected from reading the docs. I thought I'll ask nevertheless because apparently #72856 was merged to 1.10.
